### PR TITLE
Using `std` for calendrical calculations if available

### PR DESCRIPTION
**Pull request by robertbastian**
This should be reviewed commit by commit

### CODE REVIEW

File: utils/calendrical_calculations/src/helpers.rs [hunk 1]
```diff
@@ -351,270 +199,7 @@ fn test_invert_angular() {
...
-            assert_eq!(x1, x2 as i32);
-            assert_eq!(x1, x3);
-            assert_eq!(y1, y2 as i32);
-            assert_eq!(y1, y3);
-        }
-    }
-}
-
-#[test]
-fn test_ceil_div() {
```
Comment (sffc) [inline utils/calendrical_calculations/src/helpers.rs:559]: Please restore tests for functions that are implemented in this file

**Comment by Manishearth** [comment]
please let's not touch this crate while it's in the midst of a major active refactor

File: utils/calendrical_calculations/src/helpers.rs (hunk 1 above)
Comment (robertbastian) [reply utils/calendrical_calculations/src/helpers.rs:559]: I'll mark PRs as ready when they are ready to review

**Comment by robertbastian** [comment]
I assumed #3985 was it. I did link this PR so you could have been aware.

**Comment by Manishearth** [comment]
No, there's the currently open PR and at least one more PR for chinese. I didn't notice this PR until today, and I'd already been working on some of those changes the day before yesterday. I've got until monday to finish this refactor without causing problems for the 1.3 timeline.

**Comment by robertbastian** [comment]
I would like this to be in 1.3 as well, this is a performance optimisation, not just a cleanup.

**Comment by Manishearth** [comment]
In what way does this improve performance?

Besides, calendar performance for the lunar calendars is orders of magnitude off of where it ought to be and while we have a plan for fixing that it won't happen in 1.3, so in my book we've already decided we do not care about 1.3 calendar performance.

Sure, this would be nice to have in 1.3; but it's by no means a blocker IMO, whereas the other one is an immovable release blocker.

**Comment by robertbastian** [comment]
Using std methods allows using architecture-specific intrinsics.

**Comment by Manishearth** [comment]
Okay, I suspect that is multiple orders of magnitude less impactful than the data work which we have already decided to not do for 1.3.

So yeah I'd be glad if this lands in time but it does not appear to be a blocker to me.

**Comment by Manishearth** [comment]
On the plus side, I finished the main part of the migration in that one PR so it's okay to work on this now based on those changes.

**Comment by Manishearth** [comment]
Feel free to land after rebasing, I am no longer doing major changes!

File: utils/calendrical_calculations/src/astronomy.rs [hunk 2]
```diff
@@ -234,27 +242,27 @@ impl Astronomical {
 
         if (2051..=2150).contains(&year_int) {
             (-20.0
-                + 32.0 * libm::pow((year_int as f64 - 1820.0) / 100.0, 2.0)
-                + 0.5628 * (2150.0 - year_int as f64))
+                + 32.0 * (((year_int - 1820) * (year_int - 1820)) as f64 / 10000.0)
```
Comment (sffc) [inline utils/calendrical_calculations/src/astronomy.rs:245]: Note: Andrew found that calling pow was faster than calling multiply in some of these cases, which I found surprising, but he has data

**Review by sffc — COMMENTED** [review]
Architecturally I'm quite happy with this. I checked some of the math changes and they look correct to me.

File: components/calendar/src/buddhist.rs [hunk 3]
```diff
@@ -226,7 +226,7 @@ fn iso_year_as_buddhist(year: i32) -> types::FormattableYear {
 
 #[cfg(test)]
 mod test {
-    use crate::rata_die::RataDie;
+    use calendrical_calculations::rata_die::RataDie;
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/calendar/src/buddhist.rs:229]: I removed the local `helpers` reexport and thought it was clearer to remove the `rata_die` one as well.

File: utils/calendrical_calculations/src/astronomy.rs (hunk 2 above)
Comment (robertbastian) [reply utils/calendrical_calculations/src/astronomy.rs:245]: do you know if he compared to multiplying floats or integers? I'd be surprised if the floating point operations here are cheaper than the integer operations and one float division.

File: components/calendar/src/types.rs [hunk 4]
```diff
@@ -562,7 +561,7 @@ impl Time {
     /// Takes a number of minutes, which could be positive or negative, and returns the Time
     /// and the day number, which could be positive or negative.
     pub(crate) fn from_minute_with_remainder_days(minute: i32) -> (Time, i32) {
-        let (extra_days, minute_in_day) = helpers::div_rem_euclid(minute, 1440);
+        let (extra_days, minute_in_day) = (minute.div_euclid(1440), minute.rem_euclid(1440));
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply components/calendar/src/types.rs:564]: They are all marked inline. Our custom implementation was actually not inline so in cross-crate calls might have inhibited optimisations.

File: utils/calendrical_calculations/Cargo.toml [hunk 5]
```diff
@@ -31,14 +31,12 @@ independent = true
 [package.metadata.docs.rs]
 all-features = true
 
-
-
 [dependencies]
-libm = { version = "0.2", default-features = false }
+libm = "0.2"
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply utils/calendrical_calculations/Cargo.toml:35]: `libm` doesn't have any default features, and no-default-features is semver wild-west

File: utils/calendrical_calculations/src/helpers.rs [hunk 6]
```diff
@@ -684,3 +332,118 @@ fn test_i64_to_saturated_i32() {
     assert_eq!(i64_to_saturated_i32(2147483649), 2147483647);
     assert_eq!(i64_to_saturated_i32(i64::MAX), i32::MAX);
 }
+
+/// Core maths
+pub(crate) trait CoreFloat {
```
Comment (Manishearth) [inline utils/calendrical_calculations/src/helpers.rs:337]: please document these traits as to how they work

**Review by Manishearth — APPROVED** [review]

**Pull request merged**